elastic/kibana
# [Alerting] Add telemetry to count number of failed or unrecognized rule tasks

Resolves https://github.com/elastic/kibana/issues/122985

## Summary

* Updates alerting telemetry task to include a daily count of  alerting tasks with `status == 'failed' | 'unrecognized'` by querying the task manager index.

## To Verify

1. Update the alerting telemetry task to run more frequently:

```
--- a/x-pack/plugins/alerting/server/usage/task.ts
+++ b/x-pack/plugins/alerting/server/usage/task.ts
@@ -139,7 +139,7 @@ export function telemetryTaskRunner(
                   avg_execution_time_per_day: dailyExecutionCounts.avgExecutionTime,
                   avg_execution_time_by_type_per_day: dailyExecutionCounts.avgExecutionTimeByType,
                 },
-                runAt: getNextMidnight(),
+                runAt: new Date(), // getNextMidnight(),
               };
             }
           )
```

2. To force a task into a `failed` status:
  * Create a custom space
  * Create a rule (or rules) in the custom space that run frequently
  * Delete the custom space
  * Query for alerting tasks in task manager and see that they should be in `status: failed`

```
GET .kibana_task_manager/_search
{
  "size": 100,
  "query": {
    "wildcard": {
      "task.taskType": {
        "value": "alerting:*"
      }
    }
  }
}
```

3. To force a task into an `unrecognized` status
  * Create some rules for a rule type that run frequently. I created some `People in Space` rules
  * Update the code so that this rule type is no longer registered with alerting

```
--- a/x-pack/examples/alerting_example/server/plugin.ts
+++ b/x-pack/examples/alerting_example/server/plugin.ts
@@ -28,7 +28,7 @@ export interface AlertingExampleDeps {
 export class AlertingExamplePlugin implements Plugin<void, void, AlertingExampleDeps> {
   public setup(core: CoreSetup, { alerting, features }: AlertingExampleDeps) {
     alerting.registerType(alwaysFiringAlert);
-    alerting.registerType(peopleInSpaceAlert);
+    // alerting.registerType(peopleInSpaceAlert);
```

  * Query for alerting tasks in task manager and see that they should be in `status: unrecognized`

4. Navigate to `https://localhost:5601/api/stats?extended=true&legacy=true` and see the following fields inside alerting telemetry:
* `alerts.count_failed_and_unrecognized_rule_tasks_per_day `
* `alerts.count_failed_and_unrecognized_rule_tasks_by_status_per_day `
* `alerts.count_failed_and_unrecognized_rule_tasks_by_status_by_type_per_day `

### Checklist
